### PR TITLE
Adding Arthimetic operations

### PR DESCRIPTION
**Pull request by mortonjt**
Let me know if I'm missing anything.

Note, I didn't think it would make sense to have reverse operations (i.e. `1+Interval(1, 2)`)
So that functionality isn't added atm.

### CODE REVIEW

File: pandas/src/interval.pyx [hunk 1]
```diff
@@ -101,6 +100,54 @@ cdef class Interval(IntervalMixin):
...
+
+        if isinstance(y, numbers.Number):
+            return Interval(self.left + y, self.right + y)
+        elif isinstance(y, Interval) and isinstance(self, numbers.Number):
+            # It doesn't make sense to have reverse addition ( i.e. 1 + (1, 2) )
+            # return Interval(y.left + self, y.right + self)
+            raise TypeError("unsupported operand type(s) for +: '%s' and '%s'" %
+                            (type(self).__name__, type(y).__name__))
+        else:
+            raise TypeError("unsupported operand type(s) for +: '%s' and '%s'" %
```
Comment (shoyer) [inline pandas/src/interval.pyx]: you can simply `return NotImplemented` to get Python to raise this default error

Comment (shoyer) [inline pandas/src/interval.pyx]: see http://docs.cython.org/src/userguide/special_methods.html#arithmetic-methods

**Comment by shoyer** [comment]
I do think it probably makes sense to define the reverse operators, at least for `+` for `*` -- these are generally commutative operators. For `/` and `-`, the result will almost always be an invalid interval, so it's OK to skip them.

**Comment by mortonjt** [comment]
Just realized that multiplication and division is broken - will fix in a bit ...

**Comment by mortonjt** [comment]
Note that this seems to only work for PY3.  Not sure what's going on with PY2 ...

**Comment by mortonjt** [comment]
That should cover it.  @shoyer, do you think the tests for PY2 and PY3 can be specified in the travis test?

Otherwise, I think this should be good to go.

File: pandas/tests/test_interval.py [hunk 2]
```diff
@@ -63,10 +63,79 @@ def test_hash(self):
...
+        with self.assertRaises(TypeError):
+            self.interval + Interval(1, 2)
+
+    def test_math_sub(self):
+        expected = Interval(-1, 0)
+        actual = self.interval - 1
+        self.assertEqual(expected, actual)
+
+        expected = Interval(-1, 0)
+        actual = 1 - self.interval
```
Comment (shoyer) [inline pandas/tests/test_interval.py]: I don't think this is right -- `[0, 1) - 1` should be `[-1, 0)`, but `1 - [0, 1)` should be `[1, 0)`, which is not a valid interval.

File: pandas/src/interval.pyx [hunk 3]
```diff
@@ -101,6 +100,46 @@ cdef class Interval(IntervalMixin):
...
+        elif isinstance(y, Interval) and isinstance(self, numbers.Number):
+            return Interval(y.left + self, y.right + self)
+        else:
+            raise NotImplemented
+
+    def __sub__(self, y):
+        if isinstance(y, numbers.Number):
+            return Interval(self.left - y, self.right - y)
+        elif isinstance(y, Interval) and isinstance(self, numbers.Number):
+            return Interval(y.left - self, y.right - self)
```
Comment (shoyer) [inline pandas/src/interval.pyx]: I think this needs to be `self - y.left, self - y.right`

File: pandas/tests/test_interval.py [hunk 4]
```diff
@@ -63,10 +63,79 @@ def test_hash(self):
...
+        self.assertEqual(expected, actual)
+
+        actual = self.interval
+        actual /= 2
+        self.assertEqual(expected, actual)
+
+        with self.assertRaises(TypeError):
+            self.interval / Interval(1, 2)
+
+def test_math_div_py2(self):
```
Comment (shoyer) [inline pandas/tests/test_interval.py]: indentation

File: pandas/tests/test_interval.py [hunk 5]
```diff
@@ -63,10 +63,79 @@ def test_hash(self):
...
+        self.assertEqual(expected, actual)
+
+        actual = self.interval
+        actual *= 2
+        self.assertEqual(expected, actual)
+
+        with self.assertRaises(TypeError):
+            self.interval * Interval(1, 2)
+
+    def test_math_div_py3(self):
```
Comment (shoyer) [inline pandas/tests/test_interval.py]: I would just use `from __future__ import division` at the top of this file and only test the Python 3 behavior.

File: pandas/src/interval.pyx (hunk 3 above)
Comment (shoyer) [inline pandas/src/interval.pyx]: that said, negating an interval is not going to create another valid interval (it will be out of order) -- so you can possibly skip this condition.

File: pandas/src/interval.pyx [hunk 6]
```diff
@@ -101,6 +100,46 @@ cdef class Interval(IntervalMixin):
...
+        else:
+            raise NotImplemented
+
+    def __mul__(self, y):
+        if isinstance(y, numbers.Number):
+            return Interval(self.left * y, self.right * y)
```
Comment (shoyer) [inline pandas/src/interval.pyx:119]: probably a good idea to include the reflexive case for multiplication, though

**Comment by shoyer** [comment]
Travis already tests both python 2 and python 3

File: pandas/src/interval.pyx (hunk 3 above)
Comment (mortonjt) [inline pandas/src/interval.pyx]: :+1:

File: pandas/src/interval.pyx (hunk 6 above)
Comment (mortonjt) [inline pandas/src/interval.pyx:119]: Done

File: pandas/tests/test_interval.py (hunk 5 above)
Comment (mortonjt) [inline pandas/tests/test_interval.py]: That's actually what I did originally.  However, I ran into this error.

``` bash
pandas/src/interval.pyx:1:31: from __future__ imports must occur at the beginning of the file
building 'pandas.lib' extension
gcc -fno-strict-aliasing -I/Users/mortonjt/miniconda2/envs/pd2/include -arch x86_64 -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -Ipandas/src/klib -Ipandas/src -I/Users/mortonjt/miniconda2/envs/pd2/lib/python2.7/site-packages/numpy/core/include -I/Users/mortonjt/miniconda2/envs/pd2/include/python2.7 -c pandas/lib.c -o build/temp.macosx-10.5-x86_64-2.7/pandas/lib.o
pandas/lib.c:1:2: error: Do not use this file, it is the result of a failed Cython compilation.
#error Do not use this file, it is the result of a failed Cython compilation.
```

And here's a relevant, unresolved [stackoverflow post](http://stackoverflow.com/questions/28031736/cython-linking-fails-due-to-from-future-import-division-must-occur-at-the)

Which is why I opted to have a separate `__div__` method to handle PY2 division.  
What are your thoughts on this?

File: pandas/tests/test_interval.py (hunk 4 above)
Comment (mortonjt) [inline pandas/tests/test_interval.py]: Done

File: pandas/tests/test_interval.py (hunk 2 above)
Comment (mortonjt) [inline pandas/tests/test_interval.py]: Ok.  I'll get rid of that test case, as suggested in the comment below.

File: pandas/tests/test_interval.py (hunk 5 above)
Comment (shoyer) [inline pandas/tests/test_interval.py]: OK, that's a bummer.

For now, let's just make sure the unit tests can pass successfully. I would not stress checking the differences between different versions of python, and would write a single test that unambiguously works with most Python 2 and 3 (by avoiding edge cases that distinguish between floor and true division).

**Comment by mortonjt** [comment]
Modified the division test case to handle both PY2 and PY3.  Let me know if I'm missing anything

**Comment by shoyer** [comment]
looks great -- thanks!
